pkg/scheduler/core/generic_scheduler.go

```diff
@@ -235,31 +235,91 @@ func (g *genericScheduler) Preempt(pod *v1.Pod, nodeLister algorithm.NodeLister,
 	if err != nil {
 		return nil, nil, nil, err
 	}
-	nodeToVictims, err := selectNodesForPreemption(pod, g.cachedNodeInfoMap, potentialNodes, g.predicates, g.predicateMetaProducer, g.schedulingQueue, pdbs)
+	nodeToVictims, err := selectNodesForPreemption(pod, g.cachedNodeInfoMap, potentialNodes,
+		g.predicates, g.predicateMetaProducer, g.schedulingQueue, pdbs)
 	if err != nil {
 		return nil, nil, nil, err
 	}
-	for len(nodeToVictims) > 0 {
-		node := pickOneNodeForPreemption(nodeToVictims)
-		if node == nil {
-			return nil, nil, nil, err
+
+	// Use extender.Filter() to check nodeToVictims and delete nodes which failed to pass: they are
+	// not eligible for preemption as they can't even run the preemptor pod.
+	nodeToVictims, err = g.filterNodeToVictimsByExtenders(pod, nodeToVictims, g.cachedNodeInfoMap)
+	if err != nil {
+		return nil, nil, nil, err
+	}
+
+	node := pickOneNodeForPreemption(nodeToVictims)
+	if node == nil {
+		return nil, nil, nil, err
+	}
+	// Lower priority pods nominated to run on this node, may no longer fit on
+	// this node. So, we should remove their nomination. Removing their
+	// nomination updates these pods and moves them to the active queue. It
+	// lets scheduler find another place for them.
+	nominatedPods := g.getLowerPriorityNominatedPods(pod, node.Name)
+
+	return node, nodeToVictims[node].pods, nominatedPods, err
+}
+
+// filterNodeToVictimsByExtenders passes given nodeToVictims to extenders and filter out nodes which can
+// not run the preemptor pod.
+// Returns: modified nodeToVictims.
+func (g *genericScheduler) filterNodeToVictimsByExtenders(
+	pod *v1.Pod,
+	nodeToVictims map[*v1.Node]*Victims,
+	nodeNameToInfo map[string]*schedulercache.NodeInfo,
+) (map[*v1.Node]*Victims, error) {
+	// If there is no extender, return nodeToVictims directly.
+	if len(g.extenders) == 0 {
+		return nodeToVictims, nil
+	}
+
+	candidateNodes := []*v1.Node{}
+
+	// Remove the victims from the corresponding nodeInfo and send nodes to the
+	// extenders for filtering.
+	// TODO(harry): actually extender.Filter() do not use nodeInfo.Pods(). We may want to remove the
+	// deep copy logic.
+	nodeNameToInfoCopy := map[string]*schedulercache.NodeInfo{}
+	for nodeName, nodeInfo := range nodeNameToInfo {
+		// We don't want to really change nodeNameToInfo, so let's create a copy of it
+		nodeNameToInfoCopy[nodeName] = nodeInfo.Clone()
+	}
+	for node, victims := range nodeToVictims {
+		for _, victim := range victims.pods {
+			nodeNameToInfoCopy[node.GetName()].RemovePod(victim)
+		}
+		candidateNodes = append(candidateNodes, nodeNameToInfo[node.GetName()].Node())
+	}
+
+	failedNodes := map[string]string{}
+
+	for _, extender := range g.extenders {
+		_, failedNodesMap, err := extender.Filter(pod, candidateNodes, nodeNameToInfoCopy)
```

Comment:
The extender may look at its current state and fail the pod because it doesn't have resources to run this pod. How does the extender know that this call is with the intent to preempt?

Reply:
No sure how can the extender Filter call fail a pod ?

Reply:
May be I am misunderstanding this code.. I didn't follow the earlier preemption code closely.

I assumed the intent is to identify Nodes where pods can be preempted, because a new Pod needs to be scheduled. extender uses Filter call to identify free Nodes that can fit this pod. Assuming all Nodes are fully loaded because of which we are trying to preempt lower priority pods, what is the extender supposed to do on receiving this Filter call? A regular Filter call will return no Free nodes, you want it to return Nodes where preemptable pods exist right?

Reply:
The basic logic is: 
1. Scheduler check preemptor pod against all nodes without caring about the extenders, the result is `nodeToVictims` 
2. Scheduler check preemptor pod against all extenders to find those nodes which can not run it. 
3. Remove those nodes from `nodeToVictims` since they can not even run the preemptor pod. 
4. Choose the best node from modified `nodeToVictims` and do preemption.

This PR implements 2), 3). The intention is to check and remove any candidate node which can't run preemptor pod.

Reply:
@resouer 

In step 2, extender cannot tell if the Filter request is with intent to preempt other pods or not. In a full cluster, it could filter out all nodes.

Reply:
Right, so it's true that preemption will not happen on extenders side in current design. i.e. If any extender reject to run preemptor pod somehow (resource not fit or other constraint), the preemption will not happen.

We decide to do so mainly due to extender lacks the knowledge of `nodeInfo` for preemption process. 

cc @bsalamat to confirm my understanding is on right track.

Reply:
Unfortunately our documentation on scheduler extenders is almost non-existent and it is unclear what kind of state the extenders are expected to have. For example, it is unclear whether extenders are expected to have their own informers and event handlers and whether they are expected or may keep their own resource accounting. If the answer to these questions is yes, then what @ravigadde said is true and both this PR as well as the current implementation of using extenders during preemption are wrong. If the answer is no and extenders are expected to rely only on the information they receive from scheduler, then this PR and also our current implementation are correct.
@ravigadde, Since you are one of the few who use extenders and you have built the extender interface, could you please clarify which one is right?

Reply:
@bsalamat scheduler extender originated out of the need to schedule based on non k8s resources. So its another scheduler running side by side with k8s scheduler. To me, it looks like preemption needs another API or tweak to the Filter API to check for hard constraints irrespective of current usage.

Reply:
@ravigadde I think the current assumption alias with your suggestion. Preemption workflow check extender's Filter() to make sure the preemptor Pod can run on this node after doing preemption. While this is based on we assume extenders do not have own cached  info.

I noticed you actually added a `nodeCacheCapable` in extender. So how about refactoring this PR (as well as preemption logic) to work with extender when `nodeCacheCapable == false` for now? This should be in the right scope and solve the current problem we have. cc @bsalamat  WDYT?

Reply:
@ravigadde, The first version of this PR that @resouer had written included a new interface for scheduler extenders.
@resouer, I hope you still have that version and can send it again (probably as a different PR) for @ravigadde to take a look and comment.

Reply:
Thanks to CI I tracked the old commit ID :) 

@ravigadde Please take a look at #58717 , it's the original version of this implementation with a new verb and assume extender can do preemption logic.